src/transformers/models/mega/modeling_mega.py

```diff
@@ -0,0 +1,2311 @@
+        p_coeff, q_coeff = self._calc_coeffs()
+        # extend the kernel to (kernel_dim X ema_projection_size X sequence_length) and
+        # multiply q by sequential ints up to the sequence length
+        vander = torch.arange(length).to(p_coeff).view(1, 1, length) * torch.log(q_coeff)
+        kernel = (p_coeff * self.b_param) * torch.exp(vander)
+        # (kernel_dim X ema_projection_size X sequence_length) -> (kernel_dim, sequence_length)
+        return torch.einsum("dnl,dn->dl", kernel, self.g_param * self.scale)
+
+    def coeffs(self):
+        if self.training:
+            return self._calc_coeffs()
+        else:
+            if self._coeffs is None:
+                self._coeffs = self._calc_coeffs()
+            return self._coeffs
+
+    def kernel(self, length: int):
+        kernel_size = length if self.truncation is None else min(self.truncation, length)
+        if self.training:
+            return self._compute_kernel(kernel_size)
+        else:
+            if self._kernel is None or self._kernel.size(-1) < kernel_size:
+                self._kernel = self._compute_kernel(kernel_size)
+            return self._kernel[..., :kernel_size]
+
+    def step(self, inputs, length, past_state=None):
+        if length == 1:
+            return self.one_step(inputs, past_state=past_state)
+
+        # (kernel_dim X ema_projection_size X 1)
+        p_coeff, q_coeff = self.coeffs()
+        # (kernel_dim X ema_projection_size X 1+sequence_length)
+        vander = torch.arange(length + 1).to(p_coeff).view(1, 1, length + 1) * torch.log(q_coeff)
+        vander = torch.exp(vander)
+        if past_state is not None:
+            # (kernel_dim X ema_projection_size X sequence_length) * (kernel_dim X ema_projection_size X 1)
+            # -> (kernel_dim X ema_projection_size X sequence_length)
+            past_ema_proj = vander[:, :, 1:] * (self.g_param * self.scale).unsqueeze(-1)
+            # past_state will be (batch_size, kernel_dim, ema_projection_size)
+            past_ema_state = torch.einsum("bdn,dnl->bdl", past_state, past_ema_proj)
+            # (kernel_dim X ema_projection_size) * (batch_size X kernel_dim X ema_projection_size)
+            # -> (batch_size X kernel_dim X ema_projection_size)
+            past_vandermonde = vander[:, :, -1] * past_state
+        else:
+            past_ema_state = None
+            past_vandermonde = None
+
+        # (kernel_dim X ema_projection_size X sequence_length)
+        vander = vander[:, :, :-1]
+        kernel = (p_coeff * self.b_param) * vander
+        kernel_proj = torch.einsum("dnl,dn->dl", kernel, self.g_param * self.scale)
+
+        kernel_fourier = torch.fft.rfft(kernel_proj.float(), n=2 * length)
+        inputs_fourier = torch.fft.rfft(inputs.float(), n=2 * length)
+        # (batch_size X kernel_dim X sequence_length)
+        out = torch.fft.irfft(inputs_fourier * kernel_fourier, n=2 * length)[..., 0:length]
```

Comment:
This is probably the `fast_convolution` algorithm, let's create a function to add some doc on where this comes from 😉